### ISSUE TITLE
Please support hostonly mode with full portability

### ISSUE THREAD

**Issue by bdrung** [dracut-ng/dracut-ng#748]
**Describe the enhancement**

In Ubuntu, initramfs-tools builds portable initrds by default. We want that user can replace hardware (e.g. the graphics card, keyboard) and still be able to boot. Taking a disk from one machine, plug it into a different machine, and boot should work out of the box.

That was the reason Ubuntu sets `hostonly=false`. Pull request #746 and issue #674 revealed that I misunderstood the `no-hostonly` mode and I had a lack of knowledge. The man page says in "Boot parameters":

> An initramfs generated without the "hostonly" mode, does not contain any system configuration files (except for some special exceptions), so the configuration has to be done on the kernel command line.

But as you can see in #746 we want to honor the system configuration files in Ubuntu.

The usage section of the man page says:

> If you want to create lighter, smaller initramfs images, you may want to specify the --hostonly or -H option. Using this option, the resulting image will contain only those dracut modules, kernel modules and filesystems, which are needed to boot this specific machine. This has the drawback, that you can’t put the disk on another controller or machine, and that you can’t switch to another root filesystem, without recreating the initramfs image. The usage of the --hostonly option is only for experts and you will have to keep the broken pieces.

As mentioned above Ubuntu wants to support putting the disk on another controller or machine.

There is `--hostonly-mode` that can be `strict` or `sloppy`.

> In "sloppy" host-only mode, extra drivers and modules will be installed, so minor hardware change won’t make the image unbootable (e.g. changed keyboard), and the image is still portable among similar hosts.

I am proposing to add another mode, e.g. `full` that will support major hardware changes. Then Ubuntu will switch to this mode.

**Comment by bdrung** [comment]
> One way forward is to invoke dracut two times.
> 
> First invocation with `hostonly=false` to compute kernel modules (with `--kernel-only`) - for "full kernel module portability".
> 
> Second invocation with `hostonly=true` (with `--no-kernel`) to compute the rest of the initrd (which now will include any system configuration files).
> 
> I personally quite excited about this usage of dracut - similar usage is covered by one of the tests - [SYSTEMD-INITRD](https://github.com/dracut-ng/dracut-ng/blob/main/test/TEST-42-SYSTEMD-INITRD/test.sh#L77)

I like to have mode that will be equivalent to that approach.

**Comment by bdrung** [comment]
The documentation of dracut is slightly contradictory:

https://dracut-ng.github.io/dracut-ng/ says in "Host v Default mode":

> **host-only mode**
> dracut will generate a smaller customized initramfs image which contains only whatever is necessary to boot based on examining the running system.
> **default mode**
> dracut will generate a larger, but more generic, initramfs image. This is important for generic kernels, or if you are switching hardware for an installed system.

dracut(8) says:
> An initramfs generated without the "hostonly" mode, does not contain any system configuration files (except for some special exceptions), so the configuration has to be done on the kernel command line.

There are two different approaches to address this:

## Option 1: generic mode

The first option would be to introduce a generic mode (--generic). This mode will exclude any system configuration files. Otherwise the system configuration files will be included. Then we will have these setup:

* default: full kernel modules with system configuration (i.e. the newly requested mode)
* --hostonly: same as currently --hostonly
* --generic: same as the current default
* --hostonly --generic: only host specific modules, but without any system configuration files. Would that we useful? We should warn about that.

## Option 2: new hostonly mode

The second option would be to introduce a hostonly mode (--hostonly-mode=full) that implements the requested new mode.

## Option 3: consolidate modes

The third option is to consolidate the different modes behind one `--mode` parameter. The `--hostonly-mode` would be dropped.

* `--mode=full`: The requested new mode
* `--mode=hostonly`: same as `--hostonly --hostonly-mode=sloppy`
* `--mode=hostonly-strict`: same as `--hostonly --hostonly-mode=strict`
* `--mode=generic`: same as the current default

@LaszloGombos @jozzsi What do you think? Which approach do you like most? Which one should I implement?

**Comment by dalto8** [comment]
We also would like something similar to what is proposed.  Ideally, we would like to ship two initrams.  One that is narrow, and one that contains as much as sensible.
1. `hostonly=yes, hostonly_mode=strict, hostonly_cmdline=no`
2. `hostonly=full, hostonly_cmdline=no`

However, I think the current options are deeply confusing and not well named or understood.  The primary cause of confusion for me is that we are using "hostonly" in a bunch of different ways.  I think that specifying the status of system configs should be in it's own, separate option.  I also think we should stop using hostonly in so many different ways.

For example, there could still be three options, just handled a bit differently.
* include_cmdline="{yes|no}" - This is identical to the former hostonly_cmdline which would be deprecated.
* mode="{generic|hostonly|strict}" - Generic includes everything, hostonly is the same as it was prior
* exclude_system_configs="{yes|no}"

To me, this would be much easier to understand intuitively.

By using entirely different names like this, we could also deprecate the old options and maintain compatibility for some time since the deprecated options could be mapped into the new ones.

**Comment by Conan-Kudo** [comment]
I don't understand what the point of a `full` hostonly mode vs a generic mode? We have a roughly even split between default hostonly and default generic among distributions shipping it. I wonder if changing hardware without reprovisioning the OS (or at least booting up to a rescue mode to regenerate the initramfs) is systemic enough that we should consider reworking how all the initramfs generation modes work.

I also feel like after reading this whole thread that I don't fully understand what the goal is or what specifically is wanted here.

**Comment by bdrung** [comment]
> I don't understand what the point of a `full` hostonly mode vs a generic mode? We have a roughly even split between default hostonly and default generic among distributions shipping it. I wonder if changing hardware without reprovisioning the OS (or at least booting up to a rescue mode to regenerate the initramfs) is systemic enough that we should consider reworking how all the initramfs generation modes work.

The difference between a `full` hostonly mode vs a generic mode is that this `full` hostonly mode would include the system configuration files.

> I also feel like after reading this whole thread that I don't fully understand what the goal is or what specifically is wanted here.

Let me repeat what I wrote as reasoning:
> We (Ubuntu) want that user can replace hardware (e.g. the graphics card, keyboard) and still be able to boot. Taking a disk from one machine, plug it into a different machine, and boot should work out of the box.

**Comment by bdrung** [comment]
I like what @dalto8 proposed in https://github.com/dracut-ng/dracut-ng/issues/748#issuecomment-2495177858. From the discussions so far, I propose following patches:

* Rename `hostonly_cmdline` to `include_cmdline`. Deprecate the former name.
* Introduce `mode="{generic|hostonly|strict}"` where generic matches `hostonly=no`, hostonly matches `hostonly=yes hostonly_mode=sloppy`, strict matches `hostonly=yes hostonly_mode=strict`
* Add `include_system_configs={yes|no}` to decouple this behavior the mode.

Implement those changes and carry them in Ubuntu first. What do you think?

**Comment by Nowa-Ammerlaan** [comment]
> What do you think?

I think your proposal makes sense, IMO this makes the behaviour more clear compared to the current `hostonly=yes/no`.

If we are changing this anyway, then perhaps now is also the right time to reconsider the defaults. I would propose that the default is `mode=generic`, `include_system_configs=yes`, `include_cmdline=no`. This I think is the configuration that most regular users want, their config files are respected and the system remains bootable if their hardware is exchanged.

**Comment by bdrung** [comment]
> For compatibility, I think the way to do it is to add a new config file here upstream for the new 'default', and have 'make install' install this new default config file and allow Fedora, openSUSE to opt out of this new default if they prefer, but we should not force distributions like Fedora or openSUSE to change to this new behavior.

We ship `dracut.conf.d/fedora.conf.example` and `dracut.conf.d/suse.conf.example`. Are these files used? In that case we could change the default and add the needed config changes to those files to keep Fedora and openSUSE unchanged.

**Comment by dalto8** [comment]
I agree that the proposed new default is probably the most sensible for the majority of users.  However, I think there is no perfect default.

We(EndeavourOS) actually produces two initrds by default.  A "thin" one and a "fat" one.  So for us, the default doesn't matter that much.

In general, I think the default is less important than having the right options available.  Distros can easily override the defaults to match their own needs.

**Comment by Conan-Kudo** [comment]
@bdrung, I still don't get this. 

A host-specific image, by definition, *cannot* be portable to another system. But regardless, cases that you're talking about with swapping graphics cards or keyboards are expected to have working automatic fallbacks even with host-specific images these days. The only case I'm aware of where this fails is with the proprietary NVIDIA driver if you've blocked nouveau in configuration.

At that point, that's on you, since you gave dracut a configuration that allows a failure mode. And even then, if you swap out to a GPU that doesn't have its driver embedded, contemporary Linux distributions should now offer SimpleDRM as the automatic basic graphics mode. _Something_ should work.

I am unsure what you are doing in Ubuntu that would result in these fallbacks to break, because the general expectation is that unless you are swapping motherboards, everything should still work _somehow_.

**Comment by Cornelicorn** [comment]
There were instances for me already, where I had problems with this. E.g. moving a physical machine to a KVM VM (missing virtio drivers, so you need a different controller, which is mostly slower) or replacing a storage controller in servers, which then uses a different driver or one time (can't remember the exact circumstances but I think it was a switch between SATA and NVME) by just mirroring to a new drive and installing that.

In those cases the machine was essentially the same, only one part of the storage system changed.

But even when the whole system is changed, I think it would be nice to have an option to have your initrd just work, but still include system specific configuration in the initrd. E.g. you might want to have your hostname and network configuration or systemd service file overrides included, but have the ability to move the hard drive between systems.

**Comment by ArrayBolt3** [comment]
> I am unsure what you are doing in Ubuntu that would result in these fallbacks to break, because the general expectation is that **unless you are swapping motherboards**, everything should still work somehow (and even changing motherboards mostly works these days too).

But therein lies the problem. Using the same installation on a variety of motherboards is extremely common (USB installations, data recovery, hardware diagnosis, QA testing, etc.). "Mostly works" = "sometimes breaks", and for a distribution that intentionally supports motherboard changes "sometimes breaks" = broken and buggy. Having to pick between "all drivers but no config files" and "config files but missing drivers" is not workable, some of us need "all drivers *and* config files". The "full" mode proposed by @bdrung should make it so it always works unless of course you intentionally do something foolish like you mention.

**Comment by Conan-Kudo** [comment]
I think where I'm coming down on this is that we need an option that omits config and make the default generic case actually include config. The current options are not intuitive.

**Comment by dalto8** [comment]
I can explain what EndeavourOS wants.  We want a narrow initrd which is already provided by hostonly strict. Then we want a "full fat" initrd which includes everything to support major hardware changes that also includes config.  This is what is missing. Sloppy doesn't go nearly far enough for this use case.

I agree that the current settings are highly unintuitive.  I proposed above some ideas above to make it more intuitive by making it more explicit.  However, ultimately, I just want the option to exist.  If it is called `hostonly full` or `generic with configs` we can make it work either way.

**Comment by ArrayBolt3** [comment]
@jozzsi Will take a closer look at this after testing https://github.com/dracut-ng/dracut-ng/pull/1238. This sounds like what we need, thank you for your help here!

**Comment by ArrayBolt3** [comment]
Buried in GRUB development right now, but I will get to this as soon as I'm done with my current work. Thank you so much again!

**Comment by bdrung** [comment]
Coming up with an "actionable" is hard: Install a system with dracut on a disk. Then move this disk into a different machine (with different mainboard, graphics card, processor with the same architecture). Then try to boot that. Moving the disk could be done between real hardware and/or virtual machines.

I like to implement what I suggested in https://github.com/dracut-ng/dracut-ng/issues/748#issuecomment-2503732833

**Comment by bdrung** [comment]
We had a BoF at DebConf 25 and came to the conclusion that Debian wants the exact same default behavior than Ubuntu: The generated initrd should include host specific configurations and users should be able to move the OS disk from one machine to another.

This means that initrd on amd64 architecture both intel and amd microcode should be included.

https://github.com/dracut-ng/dracut-ng/pull/1351#issuecomment-2908750506 indicates that the changes to the sloppy mode increases the initrd significantly.

So we should come to an conclusion: Do we want to increase the hostonly sloppy mode to satisfy Debian/Ubuntu or should we have two separate modes:

one mode for the previous hostonly sloppy mode to match the documentation:

> so minor hardware change won’t make the image unbootable (e.g. changed keyboard)

and one for you can move the OS disk to a different machine as long as the architecture matches.

Do you agree? Or is there no use case for the first case?

**Comment by aafeijoo-suse** [comment]
> [#1351 (comment)](https://github.com/dracut-ng/dracut-ng/pull/1351#issuecomment-2908750506) indicates that the changes to the sloppy mode increases the initrd significantly.

FYI, also seen in Fedora rawhide with the recent upgrade to 107 (CC @pvalena):

- 105:

```
root@fedora:~$ dracut --version
dracut 105-3.fc43
root@fedora:~$ ls -lahtr /boot/initramfs-*
-rw-------. 1 root root 166M Jan 10  2025 /boot/initramfs-0-rescue-80fa16c3e1b44360be5f2faa6a076389.img
-rw-------. 1 root root  25M Jan 10  2025 /boot/initramfs-6.13.0-0.rc6.20250108git09a0fa92e5b4.50.fc42.x86_64.img
```

- 107:

```
root@fedora:~# dracut --version
dracut 107-5.fc43
root@fedora:~# ls -lahtr /boot/initramfs-*
-rw-------. 1 root root 254M Jul 31 09:32 /boot/initramfs-0-rescue-f1fcf89de51a442691cc7d688cfef2f4.img
-rw-------. 1 root root  59M Jul 31 09:33 /boot/initramfs-6.17.0-0.rc0.250730g4b290aae788e.3.fc43.x86_64.img
```

**Comment by aafeijoo-suse** [comment]
> > Thank goodness we didn't switch to 107 in openSUSE now that we promote grub2-bls/systemd-boot + snapshots with a small ESP.
> 
> This was an unnecessary statement to say. It's aggressive and rude. Please refrain from being antagonistic to us.

@Conan-Kudo ok, maybe it was unnecessary, but I didn't mean to be aggressive or rude, I just wanted to express that we would have the same problems that Fedora is facing now. That said, I'm sorry if anyone felt offended by this comment. We evaluated the switch to 107 in May, that was the plan, but then we [detected this problem at that time](https://github.com/dracut-ng/dracut-ng/pull/1351#issuecomment-2908750506), so we postponed it until it was resolved, because for us this is a blocker.

**Comment by bdrung** [comment]
> [@jozzsi](https://github.com/jozzsi) [@bdrung](https://github.com/bdrung) [@aafeijoo-suse](https://github.com/aafeijoo-suse) A reminder to everyone: distribution integration problems are _our_ problems. The fact that [@cmurf](https://github.com/cmurf) is surfacing it now does not change that it is our fault and our problem to resolve. We created a massive regression that was not communicated well and not okay for anyone. That fundamentally makes those dracut releases _**broken**_.

I fully agree. The noticable increased size for `hostonly=sloppy` is a regression that we need to fix. Pushing those affected distributions to use `hostonly=strict` is not the right solution IMO.

> [@bdrung](https://github.com/bdrung) This experience continues to reinforce that I don't like this direction for the hostonly mode, and I continue to think this was a bad idea. But now we have this, we need to change the default behaviors to match what dracut has promised to distributions for over a decade and get things back on track.
> 
> I will also continue to recommend Ubuntu not use the sloppy mode as it's a bad idea, but that's your call and not mine.

My initial proposal was to introduce an additional hostonly mode for full portability (to match the same functionality that initramfs-tools provides) and let Debian/Ubuntu use this new mode. As the whole discussion here shows is that making the hostonly sloppy mode to satisfy all users is not possible.

So let me (re)state my proposal:

1. Trim down `hostonly=sloppy` to a size it had before.

2. Introduce a `hostonly=portable` mode or generic + include system configs.

Then we will have 4 modes:

* `hostonly=strict`: Minimal; only what is needed for this machine; e.g. used by kdump

* `hostonly=sloppy`: contains everything `strict` contains; users can replace keyboards, mice; you can’t put the disk on another controller or machine; recommended to keep a copy of a general purpose image

* `hostonly=portable`: contains everything `sloppy` contains; users can put the disk on another controller or machine (that's what Debian/Ubuntu wants as default); user can install on USB stick (wanted by Kicksecure); users can move installations from bare metal to VMs and the other way around; no copy of a general purpose image needed

* generic aka `hostonly=no`: similar to `hostonly=portable` but without any host-specific configs

Then Fedora and other can stay on `hostonly=sloppy` and Debian, Ubuntu, Kicksecure can move to `hostonly=portable`.

@jozzsi, @Conan-Kudo, @LaszloGombos what do you think about this proposal?

**Comment by bdrung** [comment]
> For the most part, this makes sense to me. However, I don't like the word "portable" because it's a lie. Can we use a different word here?

Can you explain why that would be a lie?

Name suggestions:

* generic (could be confused with the generic mode)
* full
* moveable
* portable (not liked by @Conan-Kudo)

**Comment by Conan-Kudo** [comment]
It's a lie because it's still host-specific, even if more stuff is being included. You want to include all the things plus host configs (which means it still won't work correctly with transplanted systems). My suggestion is "expanded", since it's expanding what is included.

**Comment by bdrung** [comment]
I am fine with using "expanded" as mode name. Can you explain what transplanted systems would not work?

**Comment by dalto8** [comment]
What is the intended difference between portable/expanded and generic?  If it is only the inclusion of the configs, couldn't we simply add an option to include the configs in the generic?

**Comment by bdrung** [comment]
> What is the intended difference between portable/expanded and generic? If it is only the inclusion of the configs, couldn't we simply add an option to include the configs in the generic?

There is probably no difference (except for the configs).

**Comment by bdrung** [comment]
> > There is probably no difference (except for the configs).
> 
> I think this is the right direction - try to avoid introducing a new state for `hostonly_mode`. Out of three 3rd party dracut modules have no concept of this new mode and we should work together to see if we can find a solution changing `dracut.sh` and not changing each module.

If we do not want to add another hostonly mode, we can introduce a separate `include_system_configs={yes|no}` option to decouple this behavior from the mode.

**Comment by devkontrol** [comment]
It would be helpful to break this issue down into concrete next steps if we actually expect additional code changes from this discussion. Right now, it mixes feature requests with discoveries of existing dracut features.

> I am proposing to add another mode, e.g. full that will support major hardware changes.

Now, Ubuntu stable has been released without this mode. What is the user feedback ?
